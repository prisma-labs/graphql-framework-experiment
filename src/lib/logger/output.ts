export type Output = {
  write: (message: string) => void
}
