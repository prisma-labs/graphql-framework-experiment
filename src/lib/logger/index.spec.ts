import * as Lo from 'lodash'
import * as Logger from './'
import * as Output from './output'

resetBeforeEachTest(process, 'env')
resetBeforeEachTest(process.stdout, 'isTTY')
resetBeforeEachTest(console, 'log')

let logger: Logger.RootLogger
let output: MockOutput.MockOutput

beforeEach(() => {
  process.env.LOG_PRETTY = 'false'
  output = MockOutput.create()
  logger = Logger.create({ output })
})

describe('name', () => {
  it('becomes the first entry in path', () => {
    Logger.create({ output, name: 'foo' }).info('bar')
    expect(output.writes[0].path).toEqual(['foo'])
  })
  it('defaults to "root"', () => {
    logger.info('bar')
    expect(output.writes[0].path).toEqual(['root'])
  })
})

describe('settings', () => {
  describe('pretty', () => {
    describe('.enabled', () => {
      it('can be disabled', () => {
        expect(
          Logger.create({ pretty: { enabled: false } }).settings.pretty.enabled
        ).toEqual(false)
      })
      it('persists across peer field changes', () => {
        const l = Logger.create({ pretty: { enabled: false } })
        l.settings({ pretty: { color: false } })
        expect(l.settings.pretty).toEqual({
          enabled: false,
          color: false,
          levelLabel: false,
        })
      })
      it('controls if logs are rendered pretty or as JSON', () => {
        logger.info('foo')
        logger.settings({ pretty: true })
        logger.settings({ pretty: true, level: 'trace' })
        logger.fatal('foo', { lib: /see/ })
        logger.error('foo', { har: { mar: 'tek' } })
        logger.warn('foo', { bleep: [1, '2', true] })
        logger.info('foo', { qux: true })
        logger.debug('foo', { foo: 'bar' })
        logger.trace('foo', { a: 1, b: 2, c: 'three' })
        expect(output.writes).toMatchSnapshot()
      })
      describe('precedence', () => {
        it('considers instnace time config first', () => {
          process.stdout.isTTY = false
          process.env.LOG_PRETTY = 'false'
          const l = Logger.create({ pretty: false })
          l.settings({ pretty: true })
          expect(l.settings.pretty.enabled).toEqual(true)
        })
        it('then considers contruction time config', () => {
          process.stdout.isTTY = false
          process.env.LOG_PRETTY = 'false'
          const l = Logger.create({ pretty: true })
          expect(l.settings.pretty.enabled).toEqual(true)
        })
        it('then considers LOG_PRETTY env var true|false (case insensitive)', () => {
          process.stdout.isTTY = false
          process.env.LOG_PRETTY = 'tRuE'
          const l = Logger.create()
          expect(l.settings.pretty.enabled).toEqual(true)
        })
        it('then defaults to process.stdout.isTTY', () => {
          delete process.env.LOG_PRETTY // pre-test logic forces false otherwise
          process.stdout.isTTY = true
          const l = Logger.create()
          expect(l.settings.pretty.enabled).toEqual(true)
        })
      })
    })

    describe('.color', () => {
      it('controls if pretty logs have color or not', () => {
        logger.settings({ pretty: { enabled: true, color: false } })
        logger.info('foo', { qux: true })
        expect(output.writes).toMatchSnapshot()
      })
      it('can be disabled', () => {
        expect(
          Logger.create({ pretty: { enabled: false, color: false } }).settings
            .pretty.color
        ).toEqual(false)
      })
      it('is true by default', () => {
        expect(
          Logger.create({ pretty: { enabled: true } }).settings.pretty
        ).toEqual({
          enabled: true,
          color: true,
          levelLabel: false,
        })
        expect(
          Logger.create({ pretty: { enabled: false } }).settings.pretty
        ).toEqual({
          enabled: false,
          color: true,
          levelLabel: false,
        })
      })
      it('persists across peer field changes', () => {
        const l = Logger.create({
          pretty: { enabled: false, color: false },
        })
        l.settings({ pretty: { enabled: true } })
        expect(l.settings.pretty).toEqual({
          enabled: true,
          color: false,
          levelLabel: false,
        })
      })
    })
    describe('.levelLabel', () => {
      it('is false by default', () => {
        expect(Logger.create().settings.pretty.levelLabel).toEqual(false)
      })
      it('can be enabled', () => {
        expect(
          Logger.create({ pretty: { levelLabel: true } }).settings.pretty
            .levelLabel
        ).toEqual(true)
        expect(
          Logger.create().settings({ pretty: { levelLabel: true } }).settings
            .pretty.levelLabel
        ).toEqual(true)
      })
      it('persists across peer field changes', () => {
        const l = Logger.create({ pretty: { levelLabel: true } })
        l.settings({ pretty: false })
        expect(l.settings.pretty.levelLabel).toBe(true)
      })
      it('controls if label is spelt out or not', () => {
        logger.settings({
          pretty: { enabled: true, levelLabel: true, color: false },
        })
        logger.fatal('foo')
        logger.error('foo')
        logger.warn('foo')
        logger.info('foo')
        logger.debug('foo')
        logger.trace('foo')
        expect(output.writes).toMatchSnapshot()
      })
    })

    describe('shorthands', () => {
      it('true means enabled true', () => {
        expect(Logger.create({ pretty: true }).settings.pretty).toEqual({
          enabled: true,
          color: true,
          levelLabel: false,
        })
      })

      it('false means enabled false', () => {
        expect(Logger.create({ pretty: false }).settings.pretty).toEqual({
          enabled: false,
          color: true,
          levelLabel: false,
        })
      })
    })
  })
})

describe('demo', () => {
  it('runs a demo with fake data and all levels active', () => {
    console.log = output.write
    logger.settings({ pretty: { enabled: true, color: false } })
    Logger.demo(logger)
    expect(output.writes).toMatchSnapshot()
    output.writes.length = 0
    logger.settings({ pretty: false })
    Logger.demo(logger)
    expect(output.writes).toMatchSnapshot()
  })
  it.todo('runs automatically when LOG_DEMO=true')
})

describe('output', () => {
  it('defaults to stdout for all levels', () => {
    const write = process.stdout.write
    ;(process.stdout.write as any) = output.write
    Logger.create({ pretty: false }).fatal('foo')
    process.stdout.write = write
    expect(output.writes).toMatchSnapshot()
  })
})

describe('level', () => {
  describe('precedence', () => {
    it('considers instance time config first', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'fatal'
      const logger = Logger.create({ level: 'fatal' })
      logger.settings({ level: 'trace' })
      expect(logger.settings.level).toEqual('trace')
    })

    it('then considers construction time config', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'fatal'
      const logger = Logger.create({ level: 'trace' })
      expect(logger.settings.level).toEqual('trace')
    })

    it('then considers LOG_LEVEL env var', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'trace'
      const logger = Logger.create()
      expect(logger.settings.level).toEqual('trace')
    })

    it('then considers NODE_ENV=production', () => {
      process.env.NODE_ENV = 'production'
      const logger = Logger.create()
      console.log(process.env.LOG_LEVEL)
      expect(logger.settings.level).toEqual('info')
    })

    it('then defaults to debug', () => {
      const logger = Logger.create()
      expect(logger.settings.level).toEqual('debug')
    })
  })

  it('logs below set level are not output', () => {
    logger.settings({ level: 'warn' }).info('foo')
    expect(output.writes).toEqual([])
  })

  it('LOG_LEVEL env var config is treated case insensitive', () => {
    process.env.NODE_ENV = 'production'
    process.env.LOG_LEVEL = 'TRACE'
    const logger = Logger.create()
    expect(logger.settings.level).toEqual('trace')
  })

  it('LOG_LEVEL env var config when invalid triggers thrown readable error', () => {
    process.env.LOG_LEVEL = 'ttrace'
    expect(() => Logger.create()).toThrowErrorMatchingInlineSnapshot(
      `"Could not parse environment variable LOG_LEVEL into LogLevel. The environment variable was: ttrace. A valid environment variable must be like: fatal, error, warn, info, debug, trace"`
    )
  })
})

describe('.<level> log methods', () => {
  it('accept an event name and optional context', () => {
    logger.info('hi', { user: { id: 1 } })
    logger.info('bye')
    expect(output.writes).toMatchSnapshot()
  })

  it('one for each log level', () => {
    logger.settings({ level: 'trace' })
    logger.fatal('hi')
    logger.error('hi')
    logger.warn('hi')
    logger.info('hi')
    logger.debug('hi')
    logger.trace('hi')
    expect(output.writes).toMatchSnapshot()
  })
})

describe('.addToContext', () => {
  it('pins context for all subsequent logs from the logger', () => {
    logger.addToContext({ user: { id: 1 } })
    logger.info('hi')
    expect(output.writes).toMatchSnapshot()
  })

  it('can be called multiple times, merging deeply', () => {
    logger.addToContext({ user: { id: 1 } })
    logger.addToContext({ user: { name: 'Jill' } })
    logger.info('hi')
    expect(output.writes).toMatchSnapshot()
  })

  it('gets deeply merged with local context', () => {
    logger.addToContext({ user: { id: 1 } })
    logger.info('hi', { user: { name: 'Jill' } })
    expect(output.writes).toMatchSnapshot()
  })

  it('local context takes prescedence over pinned context', () => {
    logger.addToContext({ user: { id: 1 } })
    logger.info('hi', { user: { id: 2 } })
    expect(output.writes).toMatchSnapshot()
  })
})

describe('.child', () => {
  it('creates a sub logger', () => {
    logger.child('tim').info('hi')
    expect(output.writes).toMatchSnapshot()
  })

  it('log output includes path field showing the logger namespacing', () => {
    logger
      .child('b')
      .child('c')
      .child('d')
      .info('foo')
    expect(output.writes[0].path).toEqual(['root', 'b', 'c', 'd'])
  })

  it('inherits context from parent', () => {
    logger
      .addToContext({ foo: 'bar' })
      .child('tim')
      .info('hi')
    expect(output.writes[0].context).toEqual({ foo: 'bar' })
  })

  it('at log time reflects the current state of parent context', () => {
    const b = logger.child('b')
    logger.addToContext({ foo: 'bar' })
    b.info('lop')
    expect(output.writes[0].context).toEqual({ foo: 'bar' })
  })

  it('at log time reflects the current state of parent context even from further up the chain', () => {
    const b = logger.child('b')
    const c = b.child('c')
    const d = c.child('d')
    logger.addToContext({ foo: 'bar' })
    d.info('lop')
    expect(output.writes[0].context).toEqual({ foo: 'bar' })
  })

  it('inherits level from parent', () => {
    expect(logger.settings.level).toBe('debug')
    logger
      .settings({ level: 'trace' })
      .child('tim')
      .trace('hi')
    // The fact that we get output for trace log from child means it honored the
    // setLevel.
    expect(output.writes).toMatchSnapshot()
  })

  it('reacts to level changes in root logger', () => {
    const b = logger.child('b')
    logger.settings({ level: 'trace' })
    b.trace('foo')
    // The fact that we get output for trace log from child means it honored the
    // setLevel.
    expect(output.writes).toMatchSnapshot()
  })

  it('is unable to change context of parent', () => {
    logger.child('b').addToContext({ foo: 'bar' })
    logger.info('qux')
    expect(output.writes[0].context).toEqual({})
  })

  it('is unable to change context of siblings', () => {
    const b1 = logger.child('b1').addToContext({ from: 'b1' })
    const b2 = logger.child('b2').addToContext({ from: 'b2' })
    const b3 = logger.child('b3').addToContext({ from: 'b3' })
    logger.addToContext({ foo: 'bar' })
    b1.info('foo')
    b2.info('foo')
    b3.info('foo')
    // All should inherit the root context
    expect(output.writes).toMatchSnapshot()
  })

  it('cannot affect level', () => {
    expect((logger.child('b') as any).setLevel).toBeUndefined()
  })
})

//
// Helpers
//

/**
 * A mock output stream useful for passing to logger and later reflecting on the
 * values that it wrote.
 */
namespace MockOutput {
  export type MockOutput = Output.Output & {
    writes: Record<string, any>[]
  }

  export function create(): MockOutput {
    const output = {
      writes: [],
      write(message: string) {
        let log: any
        try {
          log = JSON.parse(message)
          log.time = 0
          log.pid = 0
        } catch (e) {
          if (e instanceof SyntaxError) {
            // assume pretty mode is on
            log = message
          }
        }
        output.writes.push(log)
      },
    } as MockOutput

    return output
  }
}

/**
 * Restore the key on given object before each test. Useful for permiting tests
 * to modify the environment and so on.
 */
function resetBeforeEachTest(object: any, key: string) {
  const orig = object[key]
  beforeEach(() => {
    if (typeof orig === 'object') {
      object[key] = Lo.cloneDeep(orig)
    } else {
      object[key] = orig
    }
  })
}
