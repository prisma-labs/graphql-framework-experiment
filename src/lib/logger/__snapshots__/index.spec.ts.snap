// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.<level> log methods accept an event name and optional context 1`] = `
Array [
  Object {
    "context": Object {
      "user": Object {
        "id": 1,
      },
    },
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "bye",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.<level> log methods one for each log level 1`] = `
Array [
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 60,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 50,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 40,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 20,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 10,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.addToContext can be called multiple times, merging deeply 1`] = `
Array [
  Object {
    "context": Object {
      "user": Object {
        "id": 1,
        "name": "Jill",
      },
    },
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.addToContext gets deeply merged with local context 1`] = `
Array [
  Object {
    "context": Object {
      "user": Object {
        "id": 1,
        "name": "Jill",
      },
    },
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.addToContext local context takes prescedence over pinned context 1`] = `
Array [
  Object {
    "context": Object {
      "user": Object {
        "id": 2,
      },
    },
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.addToContext pins context for all subsequent logs from the logger 1`] = `
Array [
  Object {
    "context": Object {
      "user": Object {
        "id": 1,
      },
    },
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.child creates a sub logger 1`] = `
Array [
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
      "tim",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.child inherits level from parent 1`] = `
Array [
  Object {
    "context": Object {},
    "event": "hi",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 10,
    "path": Array [
      "root",
      "tim",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.child is unable to change context of siblings 1`] = `
Array [
  Object {
    "context": Object {
      "foo": "bar",
      "from": "b1",
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
      "b1",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "foo": "bar",
      "from": "b2",
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
      "b2",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "foo": "bar",
      "from": "b3",
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
      "b3",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`.child reacts to level changes in root logger 1`] = `
Array [
  Object {
    "context": Object {},
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 10,
    "path": Array [
      "root",
      "b",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`demo runs a demo with fake data and all levels active 1`] = `
Array [
  undefined,
  "-----------",
  "LOGGER DEMO",
  "-----------",
  "‚úï root:foo  --  lib: /see/
",
  "‚ñ† root:foo  --  har: { mar: 'tek' }  jar: [
  1, 2, 3, 4, 4, 5, 6,
  6, 7, 9, 1, 2, 4, 5,
  6, 7, 3, 6, 5, 4
]  kio: [Object: null prototype] [foo] {}
",
  "‚ñ≤ root:foo  --  bleep: [ 1, '2', true ]
",
  "‚óè root:foo 
",
  "‚óã root:foo  --  results: [
  { userId: 1, id: 1, title: 'delectus aut autem', completed: false },
  {
    userId: 1,
    id: 2,
    title: 'quis ut nam facilis et officia qui',
    completed: false
  },
  { userId: 1, id: 3, title: 'fugiat veniam minus', completed: false },
  { userId: 1, id: 4, title: 'et porro tempora', completed: true },
  {
    userId: 1,
    id: 5,
    title: 'laboriosam mollitia et enim quasi adipisci quia provident illum',
    completed: false
  }
]  tri: 'wiz'  on: false
",
  "‚óã root:foo  --  foo: 'bar'
",
  "‚Äî root:foo  --  a: 1  b: 2  c: 'three'
",
  "-----------",
  undefined,
]
`;

exports[`demo runs a demo with fake data and all levels active 2`] = `
Array [
  undefined,
  "-----------",
  "LOGGER DEMO",
  "-----------",
  Object {
    "context": Object {
      "lib": Object {},
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 60,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "har": Object {
        "mar": "tek",
      },
      "jar": Array [
        1,
        2,
        3,
        4,
        4,
        5,
        6,
        6,
        7,
        9,
        1,
        2,
        4,
        5,
        6,
        7,
        3,
        6,
        5,
        4,
      ],
      "kio": Object {},
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 50,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "bleep": Array [
        1,
        "2",
        true,
      ],
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 40,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {},
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "on": false,
      "results": Array [
        Object {
          "completed": false,
          "id": 1,
          "title": "delectus aut autem",
          "userId": 1,
        },
        Object {
          "completed": false,
          "id": 2,
          "title": "quis ut nam facilis et officia qui",
          "userId": 1,
        },
        Object {
          "completed": false,
          "id": 3,
          "title": "fugiat veniam minus",
          "userId": 1,
        },
        Object {
          "completed": true,
          "id": 4,
          "title": "et porro tempora",
          "userId": 1,
        },
        Object {
          "completed": false,
          "id": 5,
          "title": "laboriosam mollitia et enim quasi adipisci quia provident illum",
          "userId": 1,
        },
      ],
      "tri": "wiz",
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 20,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "foo": "bar",
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 20,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  Object {
    "context": Object {
      "a": 1,
      "b": 2,
      "c": "three",
    },
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 10,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  "-----------",
  undefined,
]
`;

exports[`output defaults to stdout for all levels 1`] = `
Array [
  Object {
    "context": Object {},
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 60,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
]
`;

exports[`settings pretty .color controls if pretty logs have color or not 1`] = `
Array [
  "‚óè root:foo  --  qux: true
",
]
`;

exports[`settings pretty .enabled controls if logs are rendered pretty or as JSON 1`] = `
Array [
  Object {
    "context": Object {},
    "event": "foo",
    "hostname": "Jasons-Prisma-Machine.local",
    "level": 30,
    "path": Array [
      "root",
    ],
    "pid": 0,
    "time": 0,
    "v": 1,
  },
  "[31m‚úï root[39m[90m:[39mfoo  [90m--[39m  [90mlib[39m[90m: [39m[31m/see/[39m
",
  "[31m‚ñ† root[39m[90m:[39mfoo  [90m--[39m  [90mhar[39m[90m: [39m{ mar: [32m'tek'[39m }
",
  "[33m‚ñ≤ root[39m[90m:[39mfoo  [90m--[39m  [90mbleep[39m[90m: [39m[ [33m1[39m, [32m'2'[39m, [33mtrue[39m ]
",
  "[32m‚óè root[39m[90m:[39mfoo  [90m--[39m  [90mqux[39m[90m: [39m[33mtrue[39m
",
  "[34m‚óã root[39m[90m:[39mfoo  [90m--[39m  [90mfoo[39m[90m: [39m[32m'bar'[39m
",
  "[35m‚Äî root[39m[90m:[39mfoo  [90m--[39m  [90ma[39m[90m: [39m[33m1[39m  [90mb[39m[90m: [39m[33m2[39m  [90mc[39m[90m: [39m[32m'three'[39m
",
]
`;

exports[`settings pretty .levelLabel controls if label is spelt out or not 1`] = `
Array [
  "‚úï fatal root:foo 
",
  "‚ñ† error root:foo 
",
  "‚ñ≤ warn  root:foo 
",
  "‚óè info  root:foo 
",
  "‚óã debug root:foo 
",
]
`;
